# Finding out what App Genres are popular in the App Store and Google Play Markets

## The Introduction

In this project I am grabbing some Google Play and App Store datasets from Kaggle.com and will be exploring the data to find out which category of app appears to be the most popular and therefore the most profitable.

I got both datasets from Kaggle.com.

Android Dataset: https://www.kaggle.com/lava18/google-play-store-apps/home

iOS Dataset:https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home

I believe both datasets are only samples of the real actual numbers in the app store and google market, but it is sufficient for this project's purposes.

I used Dataquest.com as a guide and reference in completeing this project and also other forums for help in certain sections. 

## The Stiuation

I am working as a data analyst for a company that builds Android apps. They occasionaly build ios apps IF the android app they made does really well. 

My job is to enable the team of developers to make data-driven decisions with respect to the kind of apps they build.

At the company, they only build apps that are free to download and install.

## Opening and Exploring the Data

First lets open up the sample datasets.

In [11]:
from csv import reader

### The Google Play data set ###
openedFile = open('googleplaystore.csv')
readFile = reader(openedFile)
android = list(readFile)
androidHeader = android[0]
android = android[1:]

### The App Store data set ###
openedFile = open('AppleStore.csv')
readFile = reader(openedFile)
ios = list(readFile)
iosHeader = ios[0]
ios = ios[1:]

Next I created a function that allows me to explore the datasets. It has four parameters 'dataset', 'start', 'end' and 'rowsAndColumns'. This function allows me to explore the data at any stage during my project. Just makes things a bit easier for me.

- 'dataset' is the inputed dataset (android or ios).
- 'start' is the index where I want the function start showing me data.
- 'end' is the index where I want the function to stop showing me data.
- 'rowsAndColumns' just gives me the option if I want to function to tell me the number of rows and columns in the dataset.

In [12]:
def exploreData(dataset, start, end, rowsAndColumns=False):
    datasetSlice = dataset[start:end]    
    for row in datasetSlice:
        print(row)
        print('\n')

    if rowsAndColumns==True:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

Let's test it out!

In [13]:
print('GOOGLE PLAY DATA')
print(androidHeader)
print('\n')
exploreData(android, 0, 0, True)

print('\n')

print('APP STORE DATA')
print(iosHeader)
print('\n')
exploreData(ios, 0, 0, True)

GOOGLE PLAY DATA
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


Number of rows: 10841
Number of columns: 13


APP STORE DATA
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


Number of rows: 7197
Number of columns: 16


Works perfectly!

## Data Cleaning

NOTE: In the discussion page on the Kaggle site, it mentions there is incorrect data inserted in the 10472 app on the Google Play dataset. So I deleted the row (as seen below)

In [14]:
print(android[10472])  # incorrect row
print('\n')
print(androidHeader)  # header
print('\n')
print(android[0])      # correct row

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


NOTE: Deleted this row as it had 19 for the Rating column which is not possible as it is out of 5.

In [16]:
print(len(android))
del android[10472]
print(len(android))

10841
10840


I must now remove dupilcate apps in the data set beacause apps like 'Instagram' have multiple entries (as seen below). This was also mentioned in a discussion thread on Kaggle.

In [17]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


As you see, its the same entry of 'Instragram' that is duplicated four times.Now I shall check to see how many duplicates there are in total.

In [18]:
duplicateApps=[]
uniqueApps=[]

for app in android:
    name = app[0]
    if name in uniqueApps:
        duplicateApps.append(name)
    else:
        uniqueApps.append(name)

print('Number of duplicated apps: ', len(duplicateApps))
print('\n')
print('Some examples of duplicate apps: ', duplicateApps[:10])

Number of duplicated apps:  1181


Some examples of duplicate apps:  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


So there are 1181 duplicate entires in the android dataset. Now I need to remove those duplicates, so let me look for a similarity between the same duplicates. Will go back to our example on 'Instagram'.

In [19]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


As you can see the fourth position data entry (which is the Reviews column) is the only variance between all four duplicates. The different numbers show that the data has been collected at different times as the higher the number, means more reviews collected. I can use this information to build a criteria for removing the duplicates. This means we should only keep the duplicate with the highest number of reviews as that is the most recent data collected and remove the other entries.

To remove the duplicates I will:

1. Create a dictionary, where each dictionary key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app.

2. Use the information stored in the dictionary and create a new data set, which will have only one entry per app (and for each app, we'll only select the entry with the highest number of reviews).

In [20]:
reviewsMax = {}

for app in android:
    name = app[0]
    nReviews = float(app[3])
    
    if name in reviewsMax and reviewsMax[name] < nReviews:
        reviewsMax[name] = nReviews
    
    elif name not in reviewsMax:
        reviewsMax[name] = nReviews
        


In a previous code cell, we found that there are 1,181 cases where an app occurs more than once, so the length of my dictionary (of unique apps) should be equal to the difference between the length of my data set and 1,181.

In [21]:
print('Expected length:', len(android) - 1181)
print('Actual length:', len(reviewsMax))

Expected length: 9659
Actual length: 9659


In [22]:
androidClean=[]
alreadyAdded=[]

for app in android:
    name = app[0]
    nReviews = float(app[3])
    
    if (nReviews == reviewsMax[name]) and (name not in alreadyAdded):
        androidClean.append(app)
        alreadyAdded.append(name)

Now let's look at the androidClean data using the exploreData function that we created earlier to see if the dataset has 9659 rows (apps), as this will mean everything worked accordinly and duplicates were removed. 

In [23]:
exploreData(androidClean,0,3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


As you can see the number of rows (apps) is 9659, which means the duplicates were removed.

Now I would like to analyse apps that are directed towards an English-speaking audience. As directed on the Kaggle site, there are apps in the dataset that are geared towards other languages. For example:

In [24]:
print(androidClean[4412][0])
print(androidClean[7940][0])

中国語 AQリスニング
لعبة تقدر تربح DZ


For my data analysis, I do not want to consider results like the ones displayed above to be considered.

Behind the scenes, each character we use in a string has a corresponding number associated with it. For instance, the corresponding number for character 'a' is 97, character 'A' is 65, and character '爱' is 29,233. We can get the corresponding number of each character using the ord() built-in function.

In [25]:
print(ord('a'))
print(ord('A'))
print(ord('爱'))

97
65
29233


The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the ASCII (American Standard Code for Information Interchange) system. Based on this number range, we can build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters.

Now I'll write a function that will check if an input is English or non-English.

In [26]:
def isEnglish(string):
    
    for character in string:
        if ord(character) > 127:
            return False
        
    return True

Let me test if this function works accordingly.

In [27]:
print(isEnglish('Instagram'))
print(isEnglish('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(isEnglish('Docs To Go™ Free Office Suite'))
print(isEnglish('Instachat 😜'))

True
False
False
False


Now I wrote a function that detects non-English app names, but we saw that the function couldn't correctly identify certain English app names like 'Docs To Go™ Free Office Suite' and 'Instachat 😜'. This is because emojis and characters like ™ fall outside the ASCII range and have corresponding numbers over 127.

In [28]:
print(ord('™'))
print(ord('😜'))

8482
128540


If I am going to use the function I've created, I'll lose useful data since many English apps will be incorrectly labeled as non-English. To minimize the impact of data loss, I'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. This means all English apps with up to three emoji or other special characters will still be labeled as English. Our filter function is still not perfect, but it should be fairly effective.

So let's make those changes now.

In [29]:
def isEnglish(string):
    asciiCount = 0
    
    for character in string:
        if ord(character) > 127:
            asciiCount += 1
    
    if asciiCount > 3:
        return False
    else:
        return True

print(isEnglish('Docs To Go™ Free Office Suite'))
print(isEnglish('Instachat 😜'))
print(isEnglish('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


As you can see, with the changes applied to the function, now the data loss won't be as great.

The function is still not perfect, and very few non-English apps might get past the filter, but this seems good enough at this point.

In [30]:
androidEnglish = []

for app in androidClean:
    if isEnglish(app[0]):
        androidEnglish.append(app)
    

In [31]:
exploreData(androidEnglish, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


We can see that we're left with 9614 Android apps, after we have removed the duplicates and the non-English apps.

Now for my analysis I only want to look at the apps that are FREE. So I need to remove apps that are not FREE.

In [32]:
freeApps = []

for app in androidEnglish:
    price = app[7]
    if price == '0':
        freeApps.append(app)

androidFinal = freeApps

print(len(androidFinal))

8864


We are left with 8864 apps (all free) in the adroid dataset after data cleaning.

So far, I spent a good amount of time on cleaning data, and:

Removed inaccurate data, duplicate app entries, non-English apps and isolated the free apps.

BUT I have only cleaned up the android dataset, now let me quickly clean up the ios dataset. This data cleaning I put put as much explaining as it is the same procedurce as for the android dataset that we just finished.

### Removing Duplicates (ios)

In [33]:
print(iosHeader)
print('\n')
exploreData(ios, 0, 1, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


Number of rows: 7197
Number of columns: 16


So we have 7197 apps in the App Store in our dataset. As you can see the columns are different compared to the android dataset.

In [34]:
duplicateIosApps=[]
uniqueIosApps=[]

for app in ios:
    name = app[1]
    if name in uniqueIosApps:
        duplicateIosApps.append(name)
    else:
        uniqueIosApps.append(name)

print('Number of duplicated apps: ', len(duplicateIosApps))
print('\n')
print('Some examples of duplicate apps: ', duplicateIosApps[:10])

Number of duplicated apps:  2


Some examples of duplicate apps:  ['Mannequin Challenge', 'VR Roller Coaster']


There are only 2 duplicates.

In [35]:
for app in ios:
    name = app[1]
    if name == 'Mannequin Challenge':
        print(app)
print(iosHeader)

['1173990889', 'Mannequin Challenge', '109705216', 'USD', '0.0', '668', '87', '3.0', '3.0', '1.4', '9+', 'Games', '37', '4', '1', '1']
['1178454060', 'Mannequin Challenge', '59572224', 'USD', '0.0', '105', '58', '4.0', '4.5', '1.0.1', '4+', 'Games', '38', '5', '1', '1']
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


As you can see, there are a few columns different between the duplicats. The 'rating_count_tot' column is the one I am going to use to filter the duplicates, the duplicate with the higher ratings I will keep and the rest I will remove. 

In [36]:
ratingCountMax = {}

for app in ios:
    name = app[1]
    nRating = float(app[5])
    
    if name in ratingCountMax and ratingCountMax[name] < nRating:
        ratingCountMax[name] = nRating
    
    elif name not in ratingCountMax:
        ratingCountMax[name] = nRating

In [37]:
print(len(ratingCountMax))

7195


In [38]:
iosClean=[]
iosAdded=[]

for app in ios:
    name = app[1]
    nRating = float(app[5])
    
    if (nRating == ratingCountMax[name]) and (name not in iosAdded):
        iosClean.append(app)
        iosAdded.append(name)

In [39]:
exploreData(iosClean,0,2, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 7195
Number of columns: 16


Duplicates removed, now I will remove the Non-English Apps.

### Removing Non-English Apps

I'll use the function we created earlier (isEnglish) to filter out the English apps.

In [40]:
iosEnglish = []

for app in iosClean:
    if isEnglish(app[1]):
        iosEnglish.append(app)

In [41]:
exploreData(iosEnglish, 0, 2, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 6181
Number of columns: 16


After removing the Non-English apps, we have 6181 apps that are English and no duplicates.

### Isolate the FREE apps in ios dataset

Now I will isolate the FREE apps.

In [43]:
freeIosApps = []

for app in iosEnglish:
    price = app[4]
    if price == '0.0':
        freeIosApps.append(app)

iosFinal = freeIosApps

print(len(iosFinal))

3220


### Results after Data Cleaning

After cleaning the data, we are left with:

Android has 8864 apps.

IOS has 3220 apps.

Now we can start our analysis.

## Data Analysis

To minimize risks and overhead, the validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because the end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets.

Let's begin the analysis by getting a sense of what are the most common genres for each market. For this, I'll need to build frequency tables for a few columns in the data sets.

We'll build two functions we can use to analyze the frequency tables:

One function to generate frequency tables that show percentages
Another function that we can use to display the percentages in a descending order

In [44]:
def freqTable(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index] 
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    tablePercentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        tablePercentages[key] = percentage
    
    return tablePercentages

I now use the newly created function freqTable and apply it inside our next function displayTable, as seen below.

In [45]:
def displayTable(dataset, index):
    table = freqTable(dataset, index)
    tableDisplay = []
    for key in table:
        keyValAsTuple = (table[key], key)
        tableDisplay.append(keyValAsTuple)

    tableSorted = sorted(tableDisplay, reverse = True)
    for entry in tableSorted:
        print(entry[1], ':', entry[0])

Now I'll use the new created functions to display the frequencies of the prime_genre Column in the iOS dataset.

In [46]:
displayTable(iosFinal,11)
#This displays the prime_genre Column in iOS dataset

Games : 58.13664596273293
Entertainment : 7.888198757763975
Photo & Video : 4.968944099378882
Education : 3.6645962732919255
Social Networking : 3.291925465838509
Shopping : 2.608695652173913
Utilities : 2.515527950310559
Sports : 2.142857142857143
Music : 2.049689440993789
Health & Fitness : 2.018633540372671
Productivity : 1.7391304347826086
Lifestyle : 1.5838509316770186
News : 1.3354037267080745
Travel : 1.2422360248447204
Finance : 1.1180124223602486
Weather : 0.8695652173913043
Food & Drink : 0.8074534161490683
Reference : 0.5590062111801243
Business : 0.5279503105590062
Book : 0.43478260869565216
Navigation : 0.18633540372670807
Medical : 0.18633540372670807
Catalogs : 0.12422360248447205


58.14% of the free English apps in the App Store appear to fall under the category of games, making games the most common genre.

The runner-up is Entertainment apps with a 7.89% of the market in the App Store and the third being Photo and Video having a 4.97% share.

This gives the impression that the App Store (the part containing free English apps) is dominated by apps that are designed for fun. This is evident from roughly 78% of the apps falling under categories such as 'Games', 'Entertainment', 'Photo & Video', 'Social Networking', 'Shopping' and 'Music'.

While apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

Let's continue by examining the Genres and Category columns of the Google Play data set (two columns which seem to be related).

In [48]:
displayTable(androidFinal,1)
#This displays the Category Column in android dataset

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

Even so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the Genres column:

In [50]:
displayTable(androidFinal,9)
#This displays the Genres Column in android dataset

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

I'll find out which genre has the most users by calcuating the average number of installs for each app genre. For the Google Play data set this will be done by using the Installs column. This information is missing in the App Store data set. As a workaround I will take the total number of user ratings (rating_count_tot) as a measure.

Let's start with calculating the average number of user ratings per app genre on the App Store. To do that, we'll need to:

- Isolate the apps of each genre.
- Sum up the user ratings for the apps of that genre.
- Divide the sum by the number of apps belonging to that genre (not by the total number of apps).

A loop inside another loop is called a nested loop. I'll use a nested loop to compute the averages we mentioned above.

In [62]:
genresIos = freqTable(iosFinal,11)

for genre in genresIos:
    total = 0
    lenGenre = 0
    for app in iosFinal:
        genreApp = app[11]
        if genreApp == genre:
            userRatings = float(app[5])
            total += userRatings
            lenGenre += 1
    averageRatings = total / lenGenre
    print(genre , ':' , averageRatings)

Shopping : 26919.690476190477
Health & Fitness : 23298.015384615384
Games : 22812.92467948718
Finance : 31467.944444444445
Navigation : 86090.33333333333
News : 21248.023255813954
Entertainment : 14029.830708661417
Productivity : 21028.410714285714
Catalogs : 4004.0
Travel : 28243.8
Music : 57326.530303030304
Sports : 23008.898550724636
Social Networking : 71548.34905660378
Food & Drink : 33333.92307692308
Education : 7003.983050847458
Business : 7491.117647058823
Utilities : 18684.456790123455
Photo & Video : 28441.54375
Medical : 612.0
Book : 39758.5
Weather : 52279.892857142855
Reference : 74942.11111111111
Lifestyle : 16485.764705882353


Now we have the total number of user ratings per genre in the App Store.

As we can see, navigation apps have the highest number of user ratings but this figured is heavily influenced by Waze and Google Maps. Which is why it's not safe to say that navigation apps should be developed by our business.

In [63]:
for app in iosFinal:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:

In [64]:
for app in iosFinal:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Now let's analyze the Google Play market a bit.

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [66]:
displayTable(androidFinal, 5) # the Installs columns

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [77]:
categoriesAndroid = freqTable(androidFinal, 1)

for category in categoriesAndroid:
    total = 0
    lenCategory = 0
    for app in androidFinal:
        appCategory = app[1]
        if appCategory == category:
            noInstalls = app[5]
            noInstalls = noInstalls.replace('+', '')
            noInstalls = noInstalls.replace(',', '')
            total += float(noInstalls)
            lenCategory += 1
    averageInstalls = total / lenCategory
    print(category,':',averageInstalls)

PHOTOGRAPHY : 17840110.40229885
DATING : 854028.8303030303
PERSONALIZATION : 5201482.6122448975
ENTERTAINMENT : 11640705.88235294
COMICS : 817657.2727272727
HEALTH_AND_FITNESS : 4188821.9853479853
TRAVEL_AND_LOCAL : 13984077.710144928
EDUCATION : 1833495.145631068
GAME : 15588015.603248259
WEATHER : 5074486.197183099
BUSINESS : 1712290.1474201474
ART_AND_DESIGN : 1986335.0877192982
FOOD_AND_DRINK : 1924897.7363636363
SOCIAL : 23253652.127118643
MEDICAL : 120550.61980830671
NEWS_AND_MAGAZINES : 9549178.467741935
AUTO_AND_VEHICLES : 647317.8170731707
PARENTING : 542603.6206896552
FINANCE : 1387692.475609756
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
SPORTS : 3638640.1428571427
FAMILY : 3695641.8198090694
HOUSE_AND_HOME : 1331540.5616438356
COMMUNICATION : 38456119.167247385
EVENTS : 253542.22222222222
PRODUCTIVITY : 16787331.344927534
VIDEO_PLAYERS : 24727872.452830188
SHOPPING : 7036877.311557789
TOOLS : 10801391.298666667
BOOKS_AND_REFERENCE : 8767811.89473684

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

In [78]:
for app in androidFinal:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times:

In [79]:
under100m = []

for app in androidFinal:
    noInstalls = app[5]
    noInstalls = noInstalls.replace(',', '')
    noInstalls = noInstalls.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(noInstalls) < 100000000):
        under100m.append(float(noInstalls))
        
sum(under100m) / len(under100m)

3603485.3884615386

We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:

In [81]:
for app in androidFinal:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [83]:
for app in androidFinal:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [84]:
for app in androidFinal:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

## Conclusion

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.